lib/engine/game/g_1848/step/buy_train.rb

```diff
@@ -68,6 +68,8 @@ def ghan_private_owned?(entity)
           end
 
           def room?(entity)
+            return true if entity.trains.count { |t| t.name != '2E' } == @game.train_limit(entity) && can_buy_2e?(entity)
```

Comment:
Couldn't this whole method be collapsed down to:

entity.trains.count { |t| t.name != '2E' } <= @game.train_limit(entity) && can_buy_2e?(entity)

if can_buy_2e? works with and without the ghan private

cost = ghan.train.price
cost -= 100 if ghan_private_owned?(entity) && @game.phase.available?(ghan_train.available_on)
cost <= entity.cash

Reply:
entity.trains.count { |t| t.name != '2E' } <= @game.train_limit(entity) && can_buy_2e?(entity):

Wouldn't it return false if a corp bought a 2e, but is not at train limit yet?

Reply:
as for the second part, reading the value of the discount from the ability leads to nicer code.
100 needs explanation, and assumes knowledge of how much the discount is.

ghan_train.price - ghan_private_ability.discount is self documenating

Reply:
yes, they don't have to apply the discount.

Reply:
I realized what you're saying. can_buy_2e forces player to have the ghan private, but that's not correct.
You can buy it without the discount while being full

Reply:
My bad, I was thinking of only the 2E case. It should be || not &&.

Reply:
i made some more changes, review and let me know if you have any other comments